source/adaptive_load/adaptive_load_controller_impl.cc

```diff
@@ -0,0 +1,444 @@
+#include <chrono>
+
+#include "envoy/common/exception.h"
+#include "envoy/config/core/v3/base.pb.h"
+
+#include "nighthawk/adaptive_load/adaptive_load_controller.h"
+#include "nighthawk/adaptive_load/metrics_plugin.h"
+#include "nighthawk/adaptive_load/scoring_function.h"
+#include "nighthawk/adaptive_load/step_controller.h"
+
+#include "external/envoy/source/common/common/logger.h"
+#include "external/envoy/source/common/common/statusor.h"
+#include "external/envoy/source/common/protobuf/protobuf.h"
+
+#include "api/adaptive_load/adaptive_load.pb.h"
+#include "api/adaptive_load/metric_spec.pb.h"
+#include "api/client/options.pb.h"
+#include "api/client/output.pb.h"
+#include "api/client/service.grpc.pb.h"
+
+#include "absl/container/flat_hash_map.h"
+#include "absl/status/status.h"
+#include "absl/strings/str_join.h"
+#include "adaptive_load/metrics_plugin_impl.h"
+#include "adaptive_load/plugin_loader.h"
+
+namespace Nighthawk {
+
+namespace {
+
+using namespace Envoy; // for Envoy::Logger
+
+using nighthawk::adaptive_load::AdaptiveLoadSessionOutput;
+using nighthawk::adaptive_load::AdaptiveLoadSessionSpec;
+using nighthawk::adaptive_load::BenchmarkResult;
+using nighthawk::adaptive_load::MetricEvaluation;
+using nighthawk::adaptive_load::MetricSpec;
+using nighthawk::adaptive_load::MetricSpecWithThreshold;
+using nighthawk::adaptive_load::ThresholdSpec;
+
+/**
+ * Runs a single benchmark using a Nighthawk Service.
+ *
+ * @param nighthawk_service_stub Nighthawk Service gRPC stub.
+ * @param command_line_options Nighthawk Service benchmark request proto generated by the
+ * StepController, without the duration set.
+ * @param duration Duration to insert into the benchmark request.
+ *
+ * @return ExecutionResponse If we reached the Nighthawk Service, this is the raw ExecutionResponse
+ * proto, containing the benchmark data or possibly an error message from Nighthawk Service; if we
+ * had trouble communicating with the Nighthawk Service, we insert the details into the
+ * |error_detail| of this proto.
+ */
+nighthawk::client::ExecutionResponse PerformNighthawkBenchmark(
+    nighthawk::client::NighthawkService::StubInterface* nighthawk_service_stub,
+    const nighthawk::client::CommandLineOptions& command_line_options,
+    const Envoy::Protobuf::Duration& duration) {
+  nighthawk::client::CommandLineOptions options = command_line_options;
+  *options.mutable_duration() = duration;
+  options.mutable_open_loop()->set_value(false);
+
+  nighthawk::client::ExecutionRequest request;
+  nighthawk::client::ExecutionResponse response;
+  *request.mutable_start_request()->mutable_options() = options;
+
+  ::grpc::ClientContext context;
+  std::shared_ptr<::grpc::ClientReaderWriterInterface<nighthawk::client::ExecutionRequest,
+                                                      nighthawk::client::ExecutionResponse>>
+      stream(nighthawk_service_stub->ExecutionStream(&context));
+
+  stream->Write(request);
```

Comment:
I think `Write()` and `WritesDone()` return a bool indicating success, should we handle those?

Reply:
Done